### PR TITLE
As part of stateless mme fixed the insertion of guti into proto map

### PR DESCRIPTION
**Pull request by rsarwad**
Title: As part of stateless mme fixed the insertion of guti into proto map
Description: While inserting guti to proto map, libprotobuf was throwing error, with message "contains invalid UTF-8 data when serializing a protocol buffer"
In order to provide UTF-8 format, converted GUTI to std::string and inserted into proto map.

Test plan: Executed s1ap sanity test suite without enabling stateless flag
On enabling stateless mme flag, tested test_attach_detach.py  and test_attach_service.py test cases

### CODE REVIEW

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 1]
```diff
@@ -105,23 +105,40 @@ void MmeNasStateConverter::proto_to_hashtable_ts(
   }
 }
 
+std::string MmeNasStateConverter::mme_app_convert_guti_to_string(
+    guti_t* guti_p) {
+  size_t len = 0;
+  len        = snprintf(
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: The length can always be 20 characters, if we use the snprintf format as
%x%x%x%x%x%x%04x%02x%08x
mme_gid is 16 bits, i.e. 4 hex chars
mme_code is 8 bits, i.e. 2 hex chars
m_tmsi is 32 bits i.e. 8 hex chars

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 2]
```diff
@@ -134,24 +151,53 @@ void MmeNasStateConverter::guti_table_to_proto(
...
+    OAILOG_DEBUG(
+        LOG_MME_APP, "guti_str:%s mme_ue_id:%d\n", guti_str.c_str(), mme_ue_id);
   }
   FREE_OBJ_HASHTABLE_KEY_ARRAY(key_array_p);
 }
 
+void MmeNasStateConverter::mme_app_convert_string_to_guti(
+    guti_t* guti_p, const std::string& guti_str) {
+  int idx         = 0;
+  std::size_t num = 1;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: Rename "num" to "bytes_to_read"

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 3]
```diff
@@ -134,24 +151,53 @@ void MmeNasStateConverter::guti_table_to_proto(
...
+#define HEX_BASE_VAL 16
+    guti_p->gummei.plmn.mcc_digit1 =
+        std::stoul(guti_str.substr(idx++, num), &num, HEX_BASE_VAL);
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: Here num is being used to determine the number of bytes to read in the substring, and also to point to the next character after the last read. Can we use different variables here?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 4]
```diff
@@ -134,24 +151,53 @@ void MmeNasStateConverter::guti_table_to_proto(
...
+    guti_p->gummei.plmn.mnc_digit2 =
+        std::stoul(guti_str.substr(idx++, num), &num, HEX_BASE_VAL);
+    guti_p->gummei.plmn.mnc_digit3 =
+        std::stoul(guti_str.substr(idx++, num), &num, HEX_BASE_VAL);
+    num = 4;
+    guti_p->gummei.mme_gid =
+        std::stoul(guti_str.substr(idx, num), &num, HEX_BASE_VAL);
+    idx += num;
+    num = 1;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: If we fix the mme_code to be written in 2 characters, then we need to read 2 characters here.

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 5]
```diff
@@ -134,24 +151,53 @@ void MmeNasStateConverter::guti_table_to_proto(
...
+    guti_p->gummei.mme_code =
+        std::stoul(guti_str.substr(idx++, num), &num, HEX_BASE_VAL);
+    num            = guti_str.size() - idx;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: Similarlly, we can read fixed 8 characters for m_tmsi

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp (hunk 3 above)
Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: Yes, previously I had used two variables, but when viewed both had same value and getting updated at same time. So thought to use same variable.

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp (hunk 1 above)
Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: MME code is of 1 byte, so total number of characters will be 19.
Here is link for reference,
https://www.netmanias.com/en/post/blog/5929/lte/lte-user-identifiers-imsi-and-guti

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp (hunk 4 above)
Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: MME code is of 1 byte

File: lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 6]
```diff
@@ -105,23 +105,35 @@ void MmeNasStateConverter::proto_to_hashtable_ts(
   }
 }
 
+char* MmeNasStateConverter::mme_app_convert_guti_to_string(guti_t* guti_p) {
+#define GUTI_STRING_LEN 20
+  char* str = (char*) calloc(1, sizeof(char) * GUTI_STRING_LEN);
+  snprintf(
+      str, GUTI_STRING_LEN, "%x%x%x%x%x%x%04x%x%08x",
```
Comment (ssanadhya) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: MME code is 8 bits (value range 0-255), which would use 2 hex chars, so it should be read as %02x. Note that we are okay reading each of mcc digits as 1 hex char as they are only using 4 bits.

Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_state_converter.cpp]: Oh! my bad, somehow did wrong calculation.
sure I will correct it

**Pull request closed without merging**